render/texture.h

```diff
@@ -1,43 +1,48 @@
 #pragma once
 
+#include <GL/glew.h>
+
 #include <string>
+#include <vector>
 
 namespace pong
 {
 
 class Texture
 {
 public:
+
+    static Texture CreateFromFile(const std::string& filePath);
+    /* Assumes the data is grayscale and stored from top left to bottom right in vector */
+    static Texture CreateFromFontCharacter(const std::vector<unsigned char>& imageData, int width, int height);
+    static Texture CreateFromSolidColor(unsigned char r, unsigned char g, unsigned char b, unsigned char a);
+
     Texture() = default;
-    explicit Texture(const std::string& filePath);
-    ~Texture();
+    Texture(const unsigned char& imageData, int width, int height, GLenum format);
+    // Copy constructors are deleted because the renderer id is used to destroy the object in opengl
+    // in the destructor. The move operator gets around this by setting the renderer id to 0.
     Texture(const Texture&) = delete;
-    Texture(Texture&&) = delete;
     Texture& operator=(const Texture&) = delete;
-    Texture& operator=(Texture&&) = delete;
+    Texture(Texture&&);
+    Texture& operator=(Texture&&);
+    ~Texture();
 
     void Bind(unsigned int slot = 0) const;
     void Unbind() const;
 
     int GetWidth() const;
     int GetHeight() const;
 
+    static std::vector<unsigned char> ConvertAlphaImageToRGBA(const std::vector<unsigned char>& imageData);
+    static void FlipImageVertically(const std::vector<unsigned char>& imageData, int width, int height, int comp, std::vector<unsigned char>& outFlippedImageData);
```

Comment:
This was me trying a couple different ways to hand back a large vector of data. The first returns a stack allocated std::vector by creating a new one in the function, fillling it with data, and std::move()'ing as the return value. I imagine that is better than straight returning by value, but I also have to imagine for large vectors (which images can be sometimes) it could still get pretty expensive?

The alternative I tried was passing in a stack variables and changing it in the function as an out variable. I find that kind of interface a little weird too though. I guess another option would be to pass back a smart point of a newly created std::vector? 
What do you think about these interfaces?

Reply:
Alright, a couple things:

> and std::move()'ing as the return value

This should never be done. The things to investigate are "return value opitimization" and "copy elision". Basically, your friend the compiler does this automatically and you can mess it up by `move`ing a return value explicitly.

> for large vectors (which images can be sometimes) it could still get pretty expensive?

Moving a vector is actually EXTREMELY cheap. :) What is a vector, under the hood? What was splArray under the hood? Really, just a pointer to some data and a size.

If you are ready to look into the absolute insanity of an STL implementation, check out clang's version [here](https://github.com/llvm-mirror/libcxx/blob/78d6a7767ed57b50122a161b91f59f19c9bd0d19/include/vector#L342-L344). If you aren't busy being blinded by underscores and instanity, you'll notice there are only 3 members here, two pointers to beginning and end, and a pointer to an allocator.

Even if your vector has 10M elements, moving those elements is 3 pointer swaps. That is why std::move is SO POWERFUL. Really insane. Instead of 10M copies, you can do 3 swaps of the datatype your machine is most optimized to swap. Crazy.


> pass back a smart point of a newly created std::vector

So one thing to realize about anything heap allocated is that you now have to concern yourself with speed of the cache. In general, stack allocated variables are WAY more performant than heap allocated variables, and this has to do with the cache. (let me know if you want some sources/educational sources on this, good things to know). Basically, anything you can put on the stack, you should. Any benefit that you get by returning a pointer in terms of that specific copy is going to be shredded by your cache getting ripped up. We also just saw that moving a big vector is basically free and is O(1). 

I would recommend returning this value on the stack, then just being mindful to pass by reference and move instead of copy !

static std::vector<unsigned char> outFlippedImageData FlipImageVertically(const std::vector<unsigned char>& imageData, int width, int height, int comp );

Reply:
Ah very interesting I figured everything gets implicitly copied when just returned like that. What a cool optimization.